models/buffer_disk.go

```diff
@@ -0,0 +1,206 @@
+	path string
+
+	batchFirst uint64 // Index of the first metric in the batch
+	batchSize  uint64 // Number of metrics currently in the batch
+
+	// Ending point of metrics read from disk on telegraf launch.
+	// Used to know whether to discard tracking metrics.
+	originalEnd uint64
+}
+
+func NewDiskBuffer(name string, path string, stats BufferStats) (*DiskBuffer, error) {
+	filePath := path + "/" + name
+	walFile, err := wal.Open(filePath, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open wal file: %w", err)
+	}
+	buf := &DiskBuffer{
+		BufferStats: stats,
+		file:        walFile,
+		path:        filePath,
+	}
+	if buf.length() > 0 {
+		buf.originalEnd = buf.writeIndex()
+	}
+	return buf, nil
+}
+
+func (b *DiskBuffer) Len() int {
+	b.Lock()
+	defer b.Unlock()
+	return b.length()
+}
+
+func (b *DiskBuffer) length() int {
+	// Special case for when the read index is zero, it must be empty (otherwise it would be >= 1)
+	if b.readIndex() == 0 {
+		return 0
+	}
+	return int(b.writeIndex() - b.readIndex())
+}
+
+// readIndex is the first index to start reading metrics from, or the head of the buffer
+func (b *DiskBuffer) readIndex() uint64 {
+	index, err := b.file.FirstIndex()
+	if err != nil {
+		panic(err) // can only occur with a corrupt wal file
+	}
+	return index
+}
+
+// writeIndex is the first index to start writing metrics to, or the tail of the buffer
+func (b *DiskBuffer) writeIndex() uint64 {
+	index, err := b.file.LastIndex()
+	if err != nil {
+		panic(err) // can only occur with a corrupt wal file
+	}
+	return index + 1
+}
+
+func (b *DiskBuffer) Add(metrics ...telegraf.Metric) int {
+	b.Lock()
+	defer b.Unlock()
+
+	dropped := 0
+	for _, m := range metrics {
+		if !b.addSingleMetric(m) {
+			dropped++
+		}
+	}
+	b.BufferSize.Set(int64(b.length()))
+	return dropped
+}
+
+func (b *DiskBuffer) addSingleMetric(m telegraf.Metric) bool {
+	data, err := metric.ToBytes(m)
+	if err != nil {
+		panic(err)
+	}
+	err = b.file.Write(b.writeIndex(), data)
+	if err == nil {
+		b.metricAdded()
+		return true
+	}
+	return false
+}
+
+func (b *DiskBuffer) Batch(batchSize int) []telegraf.Metric {
+	b.Lock()
+	defer b.Unlock()
+
+	if b.length() == 0 {
+		// no metrics in the wal file, so return an empty array
+		return []telegraf.Metric{}
+	}
+	b.batchFirst = b.readIndex()
+	var metrics []telegraf.Metric
+
+	b.batchSize = 0
+	readIndex := b.batchFirst
+	endIndex := b.writeIndex()
+	for batchSize > 0 && readIndex < endIndex {
+		data, err := b.file.Read(readIndex)
+		if err != nil {
+			panic(err)
+		}
+		readIndex++
+
+		m, err := metric.FromBytes(data)
+		if errors.Is(err, metric.ErrSkipTracking) {
+			// could not look up tracking information for metric, skip
+			continue
+		}
+		if err != nil {
+			// non-recoverable error in deserialization, abort
+			panic(err)
+		}
+		if _, ok := m.(telegraf.TrackingMetric); ok && readIndex < b.originalEnd {
+			// tracking metric left over from previous instance, skip
+			continue
+		}
```

Comment:
For the two continue cases, I would expect the tracking information to be stripped, but the metric to still be added to the batch, otherwise all "old" tracking metrics will be swallowed and never reach the sink... What do you think?

Reply:
from meeting: please provide overall comment for both continue statements to explain why we are dropping tracking metrics for future readers/editors of this.